File: packages/native_crypto_android/android/src/main/kotlin/fr/pointcheval/native_crypto_android/utils/Task.kt
package fr.pointcheval.native_crypto_android.utils

class Task<T>(private var task: () -> T) {

    private var successful = false
    private var result: T? = null
    private var exception: Exception? = null

    fun isSuccessful(): Boolean {
        return successful
    }

    fun getResult(): T {
        if (successful && result != null) {
            return result!!
        } else {
            throw Exception("No result found!")
        }
    }

    fun getException(): Exception {
        if (exception != null) {
            return exception!!
        } else {
            throw Exception("No exception found!")
        }
    }

    fun call() {
        try {
            result = task()
            exception = null
            successful = true
        } catch (e: Exception) {
            exception = e
            result = null
            successful = false
        }
    }

    fun finalize(callback: (task: Task<T>) -> Unit) {
        callback(this)
    }
}

File: packages/native_crypto_ios/ios/Classes/utils/Task.swift
//
//  Task.swift
//  native_crypto_ios
//
//  Created by Hugo Pointcheval on 25/05/2022.
//

import Foundation

class Task<T> {
    
    var task: () throws -> T
    private var successful: Bool = false
    private var result: T? = nil
    private var exception: Error? = nil
    
    init(task: @escaping () throws -> T) {
        self.task = task
    }
    
    func isSuccessful() -> Bool {
        return successful
    }
    
    func getResult() -> T? {
        return result
    }
    
    func getException() -> Error {
        if (exception != nil) {
            return exception!
        } else {
            return NativeCryptoError.exceptionError
        }
    }
    
    func call() {
        do {
            result = try task()
            exception = nil
            successful = true
        } catch {
            exception = error
            result = nil
            successful = false
        }
    }
    
    func finalize(callback: (_ task: Task<T>) -> Void) {
        callback(self)
    }
}
